verbatimModuleSyntax w/transpileOnly raises diagnostic error

### Search Terms

isolatedModules, verbatimModuleSyntax

### Expected Behavior

Ts-node should work with `verbatimModuleSyntax: true`.

### Actual Behavior

With the new config flag, `ts-node` fails to load any TS files with the following error:

```
/home/jamesernator/projects/playground/node_modules/ts-node/src/index.ts:859
    return new TSError(diagnosticText, diagnosticCodes, diagnostics);
           ^
TSError: ⨯ Unable to compile TypeScript:
error TS5104: Option 'isolatedModules' is redundant and cannot be specified with option 'verbatimModuleSyntax'.

    at createTSError (/home/jamesernator/projects/playground/node_modules/ts-node/src/index.ts:859:12)
    at reportTSError (/home/jamesernator/projects/playground/node_modules/ts-node/src/index.ts:863:19)
    at /home/jamesernator/projects/playground/node_modules/ts-node/src/index.ts:1379:34
    at Object.compile (/home/jamesernator/projects/playground/node_modules/ts-node/src/index.ts:1453:13)
    at transformSource (/home/jamesernator/projects/playground/node_modules/ts-node/src/esm.ts:400:37)
    at /home/jamesernator/projects/playground/node_modules/ts-node/src/esm.ts:278:53
    at async addShortCircuitFlag (/home/jamesernator/projects/playground/node_modules/ts-node/src/esm.ts:409:15)
    at async nextLoad (node:internal/modules/esm/loader:163:22)
    at async ESMLoader.load (node:internal/modules/esm/loader:597:20)
    at async ESMLoader.moduleProvider (node:internal/modules/esm/loader:449:11) {
  diagnosticCodes: [ 5104 ]
}
```

### Steps to reproduce the problem

Run `node --loader ts-node/esm ./test.ts` (where `test.ts`) is any ts file when `verbatimModuleSyntax: true` is set.

### Minimal reproduction

https://gist.github.com/Jamesernator/5c4598178a4f4c3ce6961aa052a48d2a

### Specifications

*ts-node v10.9.1
* node v19.7.0
* compiler v5.0.2
* tsconfig.json, if you're using one:
```
{
    "include": [
        "./**/*.ts",
        "./**/*.mts",
        "./**/*.cts",
        "./**/*.json"
    ],
    "exclude": [
        "./**/*.d.ts",
        "./**/*.d.mts",
        "./**/*.d.cts"
    ],
    "compilerOptions": {
        "resolveJsonModule": true,
        "composite": true,
        "declaration": true,
        "declarationMap": true,
        "downlevelIteration": true,
        "module": "NodeNext",
        "moduleResolution": "nodenext",
        "rootDir": "./",
        "sourceMap": true,
        "target": "ES2022",
        "newLine": "lf",
        "useDefineForClassFields": true,
        "strict": true,
        "alwaysStrict": true,
        "exactOptionalPropertyTypes": true,
        "noImplicitOverride": true,
        "noUncheckedIndexedAccess": true,
        "forceConsistentCasingInFileNames": true,
        "verbatimModuleSyntax": true
    },
    "ts-node": {
        "transpileOnly": true   
    }
}
```
* package.json
```
{
    "type": "module",
    "dependencies": {
        "ts-node": "^10.9.1",
        "typescript": "^5.0.2"  
    }
}
```

--- a/src/test/helpers/version-checks.ts
+++ b/src/test/helpers/version-checks.ts
@@ -40,6 +40,8 @@ export const tsSupportsAllowImportingTsExtensions = semver.gte(
 );
 // TS 5.0 adds ability for tsconfig to `"extends": []` an array of configs
 export const tsSupportsExtendsArray = semver.gte(ts.version, '4.999.999');
+// TS 5.0 adds verbatimModuleSyntax
+export const tsSupportsVerbatimModuleSyntax = semver.gte(ts.version, '5.0.0');
 // Relevant when @tsconfig/bases refers to es2021 and we run tests against
 // old TS versions.
 export const tsSupportsEs2021 = semver.gte(ts.version, '4.3.0');

--- a/src/test/index.spec.ts
+++ b/src/test/index.spec.ts
@@ -305,28 +305,6 @@ test.suite('ts-node', (test) => {
       );
     });
 
-    test('should support transpile only mode', async () => {
-      const r = await exec(
-        `${CMD_TS_NODE_WITH_PROJECT_FLAG} --transpile-only -pe "x"`
-      );
-      if (r.err === null) {
-        throw new Error('Command was expected to fail, but it succeeded.');
-      }
-
-      expect(r.err.message).toMatch('ReferenceError: x is not defined');
-    });
-
-    test('should throw error even in transpileOnly mode', async () => {
-      const r = await exec(
-        `${CMD_TS_NODE_WITH_PROJECT_FLAG} --transpile-only -pe "console."`
-      );
-      if (r.err === null) {
-        throw new Error('Command was expected to fail, but it succeeded.');
-      }
-
-      expect(r.err.message).toMatch('error TS1003: Identifier expected');
-    });
-
     for (const flavor of [
       '--transpiler ts-node/transpilers/swc transpile-only-swc',
       '--transpiler ts-node/transpilers/swc-experimental transpile-only-swc',

--- a/src/test/transpile-only.spec.ts
+++ b/src/test/transpile-only.spec.ts
@@ -1,0 +1,51 @@
+import { createExec } from './exec-helpers';
+import { ctxTsNode, tsSupportsVerbatimModuleSyntax } from './helpers';
+import { CMD_TS_NODE_WITH_PROJECT_FLAG } from './helpers/command-lines';
+import { TEST_DIR } from './helpers/paths';
+import { expect, context } from './testlib';
+
+const test = context(ctxTsNode);
+
+const exec = createExec({
+  cwd: TEST_DIR,
+});
+
+test('should support transpile only mode', async () => {
+  const r = await exec(
+    `${CMD_TS_NODE_WITH_PROJECT_FLAG} --transpile-only -pe "x"`
+  );
+  if (r.err === null) {
+    throw new Error('Command was expected to fail, but it succeeded.');
+  }
+
+  expect(r.err.message).toMatch('ReferenceError: x is not defined');
+});
+
+test('should throw error even in transpileOnly mode', async () => {
+  const r = await exec(
+    `${CMD_TS_NODE_WITH_PROJECT_FLAG} --transpile-only -pe "console."`
+  );
+  if (r.err === null) {
+    throw new Error('Command was expected to fail, but it succeeded.');
+  }
+
+  expect(r.err.message).toMatch('error TS1003: Identifier expected');
+});
+
+test.suite(
+  'verbatimModuleSyntax w/transpileOnly should not raise configuration diagnostic',
+  (test) => {
+    test.if(tsSupportsVerbatimModuleSyntax);
+    test('test', async (t) => {
+      // Mixing verbatimModuleSyntax w/transpileOnly
+      // https://github.com/TypeStrong/ts-node/issues/1971
+      // We should *not* get:
+      // "error TS5104: Option 'isolatedModules' is redundant and cannot be specified with option 'verbatimModuleSyntax'."
+      const service = t.context.tsNodeUnderTest.create({
+        transpileOnly: true,
+        compilerOptions: { verbatimModuleSyntax: true },
+      });
+      service.compile('const foo: string = 123', 'module.ts');
+    });
+  }
+);

--- a/src/ts-transpile-module.ts
+++ b/src/ts-transpile-module.ts
@@ -1,3 +1,6 @@
+// Derived from
+// https://github.com/microsoft/TypeScript/blob/ae1b3db8ceaae7e93bddffa1eed26309068249d7/src/services/transpile.ts
+
 import type {
   CompilerHost,
   CompilerOptions,
@@ -7,6 +10,12 @@ import type {
   TranspileOutput,
 } from 'typescript';
 import type { TSCommon } from './ts-compiler-types';
+
+const optionsRedundantWithVerbatimModuleSyntax = new Set([
+  'isolatedModules',
+  'preserveValueImports',
+  'importsNotUsedAsValues',
+]);
 
 /** @internal */
 export function createTsTranspileModule(
@@ -29,7 +38,6 @@ export function createTsTranspileModule(
     Debug,
     toPath,
     getSetExternalModuleIndicator,
-    getEntries,
     addRange,
     hasProperty,
     getEmitScriptTarget,
@@ -54,6 +62,14 @@ export function createTsTranspileModule(
   }
 
   for (const option of transpileOptionValueCompilerOptions) {
+    // Do not set redundant config options if `verbatimModuleSyntax` was supplied.
+    if (
+      options.verbatimModuleSyntax &&
+      optionsRedundantWithVerbatimModuleSyntax.has(option.name)
+    ) {
+      continue;
+    }
+
     options[option.name] = option.transpileOptionValue;
   }
 
@@ -109,12 +125,18 @@ export function createTsTranspileModule(
 
   return transpileModule;
 
+  /*
+   * This function will compile source text from 'input' argument using specified compiler options.
+   * If not options are provided - it will use a set of default compiler options.
+   * Extra compiler options that will unconditionally be used by this function are:
+   * - isolatedModules = true
+   * - allowNonTsExtensions = true
+   * - noLib = true
+   * - noResolve = true
+   */
   function transpileModule(
     input: string,
-    transpileOptions2: Pick<
-      TranspileOptions,
-      'fileName' | 'moduleName' | 'renamedDependencies'
-    >,
+    transpileOptions2: TranspileOptions,
     packageJsonType: 'module' | 'commonjs' = 'commonjs'
   ): TranspileOutput {
     // if jsx is specified then treat file as .tsx
@@ -142,7 +164,7 @@ export function createTsTranspileModule(
 
     if (transpileOptions2.renamedDependencies) {
       (sourceFile as any).renamedDependencies = new Map(
-        getEntries(transpileOptions2.renamedDependencies)
+        Object.entries(transpileOptions2.renamedDependencies)
       );
     }
 

